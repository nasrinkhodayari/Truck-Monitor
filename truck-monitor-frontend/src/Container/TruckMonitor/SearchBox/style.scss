@import "../../../Assets/scss/variables.scss";

.search-box-row {
    background-color: #fff;
    position: relative;
    z-index: 1;
    box-shadow: 3px 6px 11px 4px #0da3fa57;
    border-bottom-right-radius: 10px;
    width: 60%;
    padding: 10px;
    padding-right: 20px;
    @media screen and (max-width: 992px) {
        div {
            margin-bottom: 5px;
        }
    }

    .arrow-icon {
        right: -8px;
        position: absolute;
        cursor: pointer;
        font-size: 25px;
        top: 50%;
        color: $light_gray;
        height: 25px;
        width: 25px;
        margin-top: -25px;
        i {
            font-size: large;
            color: $light_sky_blue;
        }
    }
}

.show-search-box-row {
    div.col-12 {
        display: grid;
    }
}
.hide-search-box-row {
    width: 20px;
    height: 58px;
    @media screen and (max-width: 992px) {
        height: 204px;
    }
    div.col-12 {
        display: none;
    }
}
