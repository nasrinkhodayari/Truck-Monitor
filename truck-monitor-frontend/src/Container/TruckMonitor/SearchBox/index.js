import React, { useState } from "react";
import Select from 'react-select';
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import AppInputField from "../../../Components/Input";
import "./style.scss";
import truckService from "../Services/truck-service";
import poiService from "../Services/poi-service";
import { flyMapCenter, markerIconDetector, removeMarker } from "../../../Utils/util";
import { icn_current_location, icn_path, icn_first_location } from "../../../Constance/map-icons";
import { GET_TRUCK } from '../../../Redux/Types/truck-types';

const SearchBox = props => {
    const { storeData, dispatch, t, addMarker, map } = props;
    const [selectedPOI, setSelectedPOI] = useState('');
    const [selectedPOIRadius, setSelectedPOIRadius] = useState('');
    const [searchBoxClass, setSearchBoxClass] = useState('show-search-box-row');
    const defaultSearchBoxClass = 'search-box-row';
    const [showSearchBox, setShowSearchBox] = useState(true);
    const [poiLabelList] = useState(poiService.getpoiLabelList(t));
    const [allPoiOption] = useState({
        label: t('View all'),
        value: poiLabelList.map(item => item.value)
    });
    const [poiByRadiusLabelList] = useState(poiService.getpoiByRadiusLabelList(t));
    const [disabledSearchPOI, setDisabledSearchPOI] = useState(true);
    const { REACT_APP_LICENSE_PLATE_LENGTH,
        REACT_APP_MAPBOX_ZOOM
    } = process.env;

    const zoomInAfterSearch = zoomLevel => {
        flyMapCenter({
            map: map,
            center: [
                storeData.TruckReducer.truck.current_lng,
                storeData.TruckReducer.truck.current_lat
            ],
            zoom: REACT_APP_MAPBOX_ZOOM - zoomLevel
        });
    };
    const searchTruck = evt => {

        let licensePlateLen = parseInt(REACT_APP_LICENSE_PLATE_LENGTH);

        if (evt.target.value.length < licensePlateLen) {
            removeMarker('truck');
            removeMarker('poi');
            removeMarker('poiRadius');
            setDisabledSearchPOI(true);
        }

        if (evt.target.value.length === licensePlateLen) {
            let licensePlate = evt.target.value;
            truckService.getTruckByLicensePlate({
                licensePlate: licensePlate,
                dispatch: dispatch
            })
                .then((resultData) => {
                    const truckData = resultData.data;
                    dispatch({ type: GET_TRUCK, truck: truckData });
                    flyMapCenter({
                        map: map,
                        center: [
                            truckData.current_lng,
                            truckData.current_lat
                        ],
                        zoom: REACT_APP_MAPBOX_ZOOM
                    });

                    addMarker({
                        map: map,
                        center: [truckData.source_lng, truckData.source_lat],
                        icon: icn_first_location,
                        markerType: 'path',
                        title: `Truck first location`
                    });

                    addMarker({
                        map: map,
                        center: [truckData.current_lng, truckData.current_lat],
                        icon: icn_current_location,
                        markerType: 'truck',
                        title: `Truck License Plate : ${licensePlate.toUpperCase()}`
                    });

                    truckData.truckRoute.map((routeItem) => {
                        return addMarker({
                            map: map,
                            center: routeItem,
                            icon: icn_path,
                            markerType: 'path',
                            title: routeItem
                        });
                    });
                })
                .catch((exeption) => {
                    console.log(exeption);
                });
            setDisabledSearchPOI(false);
        }
    };
    const findPOI = poi => {
        setSelectedPOI(poi);
        removeMarker('poi');
        removeMarker('poiRadius');
        poiService.getNearestPois({
            poiTypes: poi,
            truck: storeData.TruckReducer.truck,
            dispatch: dispatch
        }).then((poisList) => {
            const { data, features } = poisList;
            let poiDataList = data.features || features;

            zoomInAfterSearch(2);

            poiDataList.map(poiVal => {
                const { properties, center } = poiVal;
                let markerIcon = markerIconDetector(properties.category.split(', '));
                return addMarker({
                    map: map,
                    center: center,
                    icon: markerIcon,
                    markerType: 'poi',
                    title: `Address: ${properties.address}`
                })
            });
        }).catch((exeption) => {
            console.log(exeption);
        });
    };
    const findPOIByRadius = radius => {
        setSelectedPOIRadius(radius);
        removeMarker('poiRadius');
        removeMarker('poi');
        poiService.getNearestPoisByRadius({
            dispatch: dispatch,
            radius: radius,
            truck: storeData.TruckReducer.truck
        }).then((poisByRadiusList) => {
            const { data, features } = poisByRadiusList;
            let poiDataList = data.features || features;
            zoomInAfterSearch(1);
            poiDataList.map(poiRadiusVal => {
                const { properties, geometry } = poiRadiusVal;
                let markerIcon = [];
                markerIcon.push((properties.type).toLocaleLowerCase());
                return addMarker({
                    map: map,
                    center: geometry.coordinates,
                    icon: markerIconDetector(markerIcon),
                    markerType: 'poiRadius',
                    title: `Name: ${properties.name || properties.category_en} -
                    Distance:${parseInt(properties.tilequery.distance)} M`
                })
            });
        }).catch((exeption) => {
            console.log(exeption);
        });
    };
    return (
        <div className={`${defaultSearchBoxClass} ${searchBoxClass}`}>
            <Row id="main">
                <Col lg={4} xs={12}><AppInputField type="text"
                    maxLength={6}
                    placeholder={t('Search by license plate')}
                    changeHandler={searchTruck} /></Col>
                <Col lg={4} xs={12}>
                    <Select
                        isDisabled={disabledSearchPOI}
                        placeholder={t('Select POI type')} value={selectedPOI} onChange={findPOI} options={[allPoiOption, ...poiLabelList]} />
                </Col>
                <Col lg={4} xs={12}>
                    <Select
                        isDisabled={disabledSearchPOI}
                        placeholder={t('Select radius')} value={selectedPOIRadius} onChange={findPOIByRadius} options={poiByRadiusLabelList} />
                </Col>
                {/* <Col lg={2} xs={12}><AppButton label={t('Apply')}></AppButton></Col> */}
            </Row>
            <div className="arrow-icon"
                onClick={() => {
                    setShowSearchBox(!showSearchBox);
                    setSearchBoxClass(searchBoxClass === 'hide-search-box-row' ? 'show-search-box-row' : 'hide-search-box-row');
                }}>
                {showSearchBox
                    ? <i className="fa fa-chevron-left" aria-hidden="true"></i>
                    : <i className="fa fa-chevron-right" aria-hidden="true"></i>
                }
            </div>
        </div>
    );
}
export default SearchBox;