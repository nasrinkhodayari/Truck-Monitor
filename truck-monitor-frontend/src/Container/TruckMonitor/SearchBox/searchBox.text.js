import React from 'react';
import { shallow } from 'enzyme';
import SearchBox from './index';

describe('App SearchBox component', () => {
    it('App searchbox renders correctly', () => {
        const wrapper = shallow(<SearchBox />);
        expect(wrapper).toMatchSnapshot();
        });
});