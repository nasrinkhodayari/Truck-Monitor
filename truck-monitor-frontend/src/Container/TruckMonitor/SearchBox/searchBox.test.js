import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider } from "react-redux";

import { render, shallow } from 'enzyme';
import SearchBox from './index';
import AppInputField from '../../../Components/Input';

import axios from 'axios';

jest.mock('axios');

const mockStore = configureMockStore([thunk]);
const storeStateMock = {
    MainReducer: { loading: true, errorMessage: 'something happen' }
};

describe('App SearchBox component', () => {
    const minProps = {
        map: '',
        mapboxgl: '',
    };
    let wrapper, store;
    store = mockStore(storeStateMock);
    beforeEach(() => {
        wrapper = shallow(
            <Provider store={store}>
                <SearchBox {...minProps} />
            </Provider>
        );
    });
    it('App searchbox renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    const maxLength = 6,
        value = 'FN67KP',
        type = 'text',
        placeholder = 'license plate',
        changeHandler = jest.fn();
    let InputFieldComponent;
    InputFieldComponent = (
        <AppInputField
            value={value}
            maxLength={maxLength}
            type={type}
            placeholder={placeholder}
            changeHandler={changeHandler}
        />);
    it('there is AppInputField', () => {
        expect(wrapper.find(InputFieldComponent))
    });

   
});