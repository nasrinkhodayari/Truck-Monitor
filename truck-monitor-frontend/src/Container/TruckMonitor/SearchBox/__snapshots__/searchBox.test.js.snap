// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App SearchBox component App searchbox renders correctly 1`] = `ShallowWrapper {}`;
